book/src/proposals/booting-from-snapshot.md

```diff
@@ -0,0 +1,222 @@
+`L_i` in tower, the validator is able to determine `is_ancestor(L_i, S_d)`.
+
+Assume the "Safety Criteria" is true, we show we can then achieve the "safety"
+condition:
+
+**Proof:** A validator wants to determine whether it can vote on `T`. `T` must
+be a descendant of `S` because the validator does not play any state for
+non-descendants of `S` when booting from the snapshot. From assuming "Safety
+Criteria" above, this means for any `L_i` we can determine `is_ancestor(L_i,
+T)`. This means we have all the tools to run the algorithm from the safety
+definition.
+
+Thus to achieve safety, we want to design the snapshotting system such that the
+"Safety Criteria" is met.
+
+
+## Implementing "Safety Criteria"
+
+The snapshot is augmented to store the last `N` ancestors of `S`. These
+ancestors are incorporated into the bank hash so they can be verified when a
+validator unpacks a snapshot. Call this set `N_Ancestors`.
+
+We now implement the "Safety Criteria" in cases:
+
+### Case 1: Calculating `is_ancestor(L_i, S_d)` when `L_i` < `S`:
+
+There are two variants depending on whether the list of `N_Ancestors` goes back
+far enough to include `L_i`. This can be established by comparing `L_i` to the
+oldest slot in `N_Ancestors`, `N_Oldest`.
+
+#### Variant 1: `L_i` >= `N_Oldest`:
+
+##### Protocol:
+
+Search the list of `N_Ancestors` ancestors of `S` to check if `L_i` is an
+ancestor of `S`.
+
+##### Proof of Correctness for Protocol:
+
+This case is equivalent to determining `is_ancestor(L_i, S)`. This is because
+`is_ancestor(L_i, S_d) <=> (is_ancestor(L_i, S) && is_ancestor(S, S_d))`, and
+`is_ancestor(S, S_d)` is true by the definition of `S_d`.
+
+Now we show the protocol is sufficient to determine `is_ancestor(L_i, S)`.
+Because `L_i + N >= S` in this variant, and `N_Ancestors` has length `N`, then
+if `L_i`, is an ancestor, it has to be a member of `N_Ancestors`, so the
+protocol is sufficient.
+
+#### Variant 2: `L_i` < `N_Oldest`:
+
+##### Protocol:
+
+Search the vote state of the validator in `S_d` for `L_i`. If `L_i` is not
+found, assume it is not an ancestor.
+
+##### Proof of Correctness for Protocol:
+
+A vote for a slot will only be placed into the validator's vote state in a bank
+descended from the slot the vote is for. Therefore, a vote will only be found
+in a vote state if the vote is for an ancestor of the slot the vote state is
+found in. This protocol may falsely determine that `L_i` is not an ancestor of
+`S_d` if the vote was never accepted by a leader, but the result will be that
+the validator will take the more cautious approach of waiting for the lockout
+of the vote to expire, so it will not be slashed due to this failure.
+
+### Case 2: Calculating `is_ancestor(L_i, S_d)` when `L_i` >= `S`:
+
+##### Protocol: 
+
+If `S_d < L_i`, return false, because an ancestor cannot have a greater slot
+number. Otherwise, Let the bank state for slot `S_d` be called `B_d`. Check
```

Comment:
[nits] Move the definition of `B_d` to [above](https://github.com/solana-labs/solana/pull/6936/files#diff-cbeba24bd9987f8ac2f4e476e496d3f7R118) for clarity (define at the first use location of `B_d`)

Reply:
The other spot that had `B_d` didn't actually use it, so I just removed it from there.